### DISCUSSION TITLE
Problem with custom color palette

### DISCUSSION THREAD

**Discussion by rubeonline** [tailwindlabs/tailwindcss#5114] [Help] [answered]
Hello togehter,

i´m using gatsby v3 with tailwind and created a custom color palette which looks like that:

```
  theme: {
    colors: {
      white: '#FFFFFF',
      black: '#000',
      orange: '#FF760D',
      active: '#4D4D4D',
      gray: {
        darkest: '#333333',
        dark: '#4D4D4D',
        DEFAULT: '#878787',
        light: '#DFDFDF',
        lightest: '#F2F2F2',
      },
    }
  },
```

Now the overwriting of colors does not work as expected: I´m using first a white bg and want to overwrite it with a gray-light:
![image](https://user-images.githubusercontent.com/5664460/127490379-98a40ff5-bc40-4fbe-aaee-36350d0aa702.png)

But the bg appears in white:
![image](https://user-images.githubusercontent.com/5664460/127490962-63dc282b-5c4f-4542-80ac-fe77404e89f6.png)

**When I´m using the default color palette it works:**
![image](https://user-images.githubusercontent.com/5664460/127491278-05b51123-c797-49a5-928b-daa635baebe8.png)

![image](https://user-images.githubusercontent.com/5664460/127491227-ce8e11a9-ec0a-4b98-bb33-129d512873c4.png)

Does anybody know whats the issue here? Is this a bug?

**Comment by simonswiss** [comment, marked as answer]
Hey!

It's definitely not a bug, and I'll try my best to explain why.

You shouldn't try to override a color with another color based on the class order in your HTML.

When you do something like this:

```html
<div class="bg-gray bg-white bg-gray-light">...</div>
```

... the background color is not going to be `gray-light` necessarily, but it's going to be **the class that is defined last in your CSS source order**.

This is how the cascade works. If two selectors have the same degree of specificity, the one defined _last_ will take precedence.

In your devtools screenshot, you can see the `bg-white` class is defined _after_ the `bg-gray-ligth` class (see character number on top right):

![image](https://user-images.githubusercontent.com/485747/127595903-91304bc1-4596-4a76-8df4-9e808c1159f1.png)

And that's why your element has a white background! Changing the order of the classes in the HTML doesn't change the underlying CSS. The order in which the utility classes are generated from the config file defines that.

Same deal with your other screenshot (from default color palette). `bg-red` is defined later, and therefore gets applied:

![image](https://user-images.githubusercontent.com/485747/127596515-9c6cb0a8-574d-4cee-95f3-7b66b61d5f8e.png)

I'm not actually sure of that (never checked!), but since you have the `white` color defined before the `gray` color in your config, I assume that the colors are generated "outside-in" (starting with nested properties (gray variants), then top-level colors.

Either way - changing the order in which you define the colors is **not the right solution**. 

You should never apply two `bg-*` classes on the same element, unless one is applied for a variant (`hover:`, `md:`, ...). If you're using the [Tailwind CSS Intellisense extension](https://marketplace.visualstudio.com/items?itemName=bradlc.vscode-tailwindcss) in vscode, it will warn you against that when you do it.

```html
<div class="bg-white hover:bg-gray-light">
```

This above will work fine, and override the background color from white to light gray _on hover_.

Hope it helps!

**Reply by rubeonline** [reply]
Hi Simon,

thanks for your reply. I just wanted to use the bg-classes for active state of my navigation. Gatbsy Link Component recognizes which route is opened and adds a class for the active state:

```
<Link
                to="/"
                className="bg-gray"
                activeClassName="bg-gray-darkest"
                onClick={this.toggle}
              >
                home
</Link>
```

So therefore i need to create a own class for that...

**Reply by simonswiss** [reply]
Hey!

I see. In this case, I'd probably use the `activeStyle` option that Gatsby Link seems to provide ([according to the docs](https://www.gatsbyjs.com/docs/reference/built-in-components/gatsby-link/#add-custom-styles-for-the-currently-active-link)).

Alternatively, if you're using `JIT`, you could add an `!` at the start of the `activeClassName` string (important flag), which will make that class _always_ win:

```diff
<Link
  to="/"
  className="bg-gray"
-  activeClassName="bg-gray-darkest"
+  activeClassName="!bg-gray-darkest"
  onClick={this.toggle}
>
  home
</Link>
```

If your link is currently active, it will display like this:

```js
<Link className="bg-gray !bg-gray-darkest">...</Link>
```

And, no matter what the order of the generated CSS bg color classes is, `!bg-gray-darkest` will always win, since this is the resulting CSS:

```css
.\!bg-gray-darkest {
  background-color: #2b2b2b !important;
}
.bg-gray {
  background-color: #8c8c8c;
}
```

And in that case, even if I have placed the `!bg-gray-darkest` first in the CSS source order, the specificity of the `!important` flag will make it take precedence.

That should do it for your use case 👍
